File: javascript/packages/engine/__tests__/engine/RulePayloadProcessor.test.ts
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { RulePayloadProcessor } from '../../src/engine/RulePayloadProcessor'
import { Payloads, Rule } from 'kraken-model'
import {
    AccessibilityPayloadResult,
    ConditionEvaluation,
    DefaultValuePayloadResult,
    RuleEvaluationResults,
    VisibilityPayloadResult,
} from 'kraken-engine-api'
import { mock } from '../mock'
import { PayloadBuilder, RulesBuilder } from 'kraken-model-builder'
import { ExecutionSession } from '../../src/engine/ExecutionSession'
import { RuleOverrideContextExtractorImpl } from '../../src/engine/results/RuleOverrideContextExtractor'
import ApplicableRuleEvaluationResult = RuleEvaluationResults.ApplicableRuleEvaluationResult
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

const payloadProcessor = new RulePayloadProcessor(mock.evaluator, new RuleOverrideContextExtractorImpl(), {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    update: () => {},
})

function execute(rule: Rule): RuleEvaluationResult {
    function config(): ExecutionSession {
        return new ExecutionSession({ currencyCd: 'USD', context: {} }, {}, 'test')
    }
    return payloadProcessor.processRule(
        {
            rule,
            dataContext: mock.data.dataContextEmpty(),
            priority: 0,
        },
        config(),
    )
}
const createRule = (payload: Payloads.Payload) =>
    new RulesBuilder()
        .setContext(mock.modelTreeJson.contexts.Policy.name)
        .setName('mock')
        .setTargetPath(mock.modelTreeJson.contexts.Policy.fields.state.name)
        .setPayload(payload)
        .build()

describe('RulePayloadProcessor', () => {
    it('should process rule evaluation instance with AssertionPayload', () => {
        const rule = createRule(PayloadBuilder.asserts().that('true'))
        expect(execute(rule)).k_toBeValidRuleResult()
    })
    it('should process rule evaluation instance with LengthPayload', () => {
        const rule = createRule(PayloadBuilder.lengthLimit().limit(111))
        expect(execute(rule)).k_toBeValidRuleResult()
    })
    it('should process rule evaluation instance with DefaultValuePayload', () => {
        const rule = createRule(PayloadBuilder.default().to("'AZ'"))
        const payloadResult = (execute(rule) as ApplicableRuleEvaluationResult).payloadResult
        expect((payloadResult as DefaultValuePayloadResult).events).toHaveLength(1)
    })
    it('should process rule evaluation instance with VisibilityPayload', () => {
        const rule = createRule(PayloadBuilder.visibility().notVisible())
        const payloadResult = (execute(rule) as ApplicableRuleEvaluationResult).payloadResult
        expect((payloadResult as VisibilityPayloadResult).visible).toBeFalsy()
    })
    it('should process rule evaluation instance with AccessibilityPayload', () => {
        const rule = createRule(PayloadBuilder.accessibility().notAccessible())
        const payloadResult = (execute(rule) as ApplicableRuleEvaluationResult).payloadResult
        expect((payloadResult as AccessibilityPayloadResult).accessible).toBeFalsy()
    })
    it('should process not applicable conditional rule evaluation instance', () => {
        const rule = createRule(PayloadBuilder.accessibility().notAccessible())
        rule.condition = {
            expression: {
                expressionString: 'false',
                expressionType: 'COMPLEX',
            },
        }
        expect(execute(rule).conditionEvaluationResult.conditionEvaluation).toBe(ConditionEvaluation.NOT_APPLICABLE)
    })
    it('should process applicable conditional rule evaluation instance', () => {
        const rule = createRule(PayloadBuilder.accessibility().notAccessible())
        rule.condition = {
            expression: {
                expressionString: 'true',
                expressionType: 'COMPLEX',
            },
        }
        expect(execute(rule).conditionEvaluationResult.conditionEvaluation).toBe(ConditionEvaluation.APPLICABLE)
    })
})

File: kraken-engine/src/test/java/kraken/runtime/engine/handlers/RulePayloadProcessorTest.java
/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime.engine.handlers;

import kraken.model.payload.PayloadType;
import kraken.model.validation.ValidationSeverity;
import kraken.runtime.engine.conditions.ConditionEvaluationResult;
import kraken.runtime.engine.conditions.RuleApplicabilityEvaluator;
import kraken.runtime.engine.context.data.DataContext;
import kraken.runtime.engine.dto.RuleEvaluationResult;
import kraken.runtime.engine.evaluation.loop.RuleEvaluationInstance;
import kraken.runtime.engine.result.reducers.validation.OverrideDependencyExtractor;
import kraken.runtime.expressions.KrakenExpressionEvaluationException;
import kraken.runtime.expressions.KrakenExpressionEvaluator;
import kraken.runtime.model.expression.CompiledExpression;
import kraken.runtime.model.rule.RuntimeRule;
import kraken.runtime.model.rule.payload.validation.AssertionPayload;
import org.junit.Before;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.when;
import static org.mockito.ArgumentMatchers.any;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

/**
 * Unit tests for {@code RulePayloadProcessorImpl} class.
 *
 * @author Tomas Dapkunas
 * @since 1.1.1
 */
public class RulePayloadProcessorTest {

    @InjectMocks
    private RulePayloadProcessorImpl testObject;

    @Mock
    private KrakenExpressionEvaluator krakenExpressionEvaluator;

    @Mock
    private RuleApplicabilityEvaluator applicabilityEvaluator;

    @Mock
    private OverrideDependencyExtractor overrideDependencyExtractor;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        when(applicabilityEvaluator.evaluateCondition(any(), any())).thenReturn(ConditionEvaluationResult.APPLICABLE);

        testObject = RulePayloadProcessorImpl.create(krakenExpressionEvaluator, applicabilityEvaluator);
    }

    @Test
    public void shouldNotBuildOverridableContextWhenExpressionEvaluationFails() {
        when(krakenExpressionEvaluator.evaluate(any(), any(), any()))
                .thenThrow(new KrakenExpressionEvaluationException("", "", "", null));

        RuleEvaluationInstance ruleEvaluationInstance = createEvaluationInstance(
                createAssertionPayloadRule(createAssertionPayload(true)), new DataContext());

        RuleEvaluationResult result = testObject.process(ruleEvaluationInstance, null);

        verify(overrideDependencyExtractor, times(0))
                .extractOverrideDependencies(any(), any());
        verify(krakenExpressionEvaluator, times(0))
                .evaluateTargetField(any(), any());

        assertTrue(result.getOverrideInfo().isOverridable());
        assertNull(result.getOverrideInfo().getOverridableRuleContextInfo());
    }

    private RuleEvaluationInstance createEvaluationInstance(RuntimeRule runtimeRule, DataContext dataContext) {
        RuleEvaluationInstance ruleEvaluationInstance = mock(RuleEvaluationInstance.class);
        when(ruleEvaluationInstance.getRule()).thenReturn(runtimeRule);
        when(ruleEvaluationInstance.getDataContext()).thenReturn(dataContext);

        return ruleEvaluationInstance;
    }

    private RuntimeRule createAssertionPayloadRule(AssertionPayload assertionPayload) {
        RuntimeRule runtimeRule = mock(RuntimeRule.class);

        when(runtimeRule.getPayload()).thenReturn(assertionPayload);
        when(runtimeRule.getContext()).thenReturn("RuleCtx");
        when(runtimeRule.getName()).thenReturn("RuleName");

        return runtimeRule;
    }

    private AssertionPayload createAssertionPayload(boolean isOverridable) {
        AssertionPayload assertionPayload = mock(AssertionPayload.class);
        CompiledExpression compiledExpression = mock(CompiledExpression.class);

        when(compiledExpression.getExpressionString()).thenReturn("");
        when(assertionPayload.getType()).thenReturn(PayloadType.ASSERTION);
        when(assertionPayload.getSeverity()).thenReturn(ValidationSeverity.critical);
        when(assertionPayload.isOverridable()).thenReturn(isOverridable);
        when(assertionPayload.getAssertionExpression()).thenReturn(compiledExpression);

        return assertionPayload;
    }

}

